textattack/commands/train_model/run_training.py

```diff
@@ -23,24 +25,253 @@
+        labels.extend(batch_labels)
+
+    model.train()
+    logits = torch.tensor(logits)
+    labels = torch.tensor(labels)
+
+    if do_regression:
+        pearson_correlation, pearson_p_value = scipy.stats.pearsonr(logits, labels)
+        return pearson_correlation
+    else:
+        preds = logits.argmax(dim=1)
+        correct = (preds == labels).sum()
+        return float(correct) / len(labels)
+
+
+def _make_directories(output_dir):
     if not os.path.exists(output_dir):
         os.makedirs(output_dir)
 
 
+def _is_writable_type(obj):
+    for ok_type in [bool, int, str, float]:
+        if isinstance(obj, ok_type):
+            return True
+    return False
+
+
 def batch_encode(tokenizer, text_list):
     if hasattr(tokenizer, "batch_encode"):
         return tokenizer.batch_encode(text_list)
     else:
         return [tokenizer.encode(text_input) for text_input in text_list]
 
 
+def _make_dataloader(tokenizer, text, labels, batch_size):
+    """
+    Create torch DataLoader from list of input text and labels.
+
+    :param tokenizer: Tokenizer to use for this text.
+    :param text: list of input text.
+    :param labels: list of corresponding labels.
+    :param batch_size: batch size (int).
+    :return: torch DataLoader for this training set.
+    """
+    text_ids = batch_encode(tokenizer, text)
+    input_ids = np.array(text_ids)
+    labels = np.array(labels)
+    data = list((ids, label) for ids, label in zip(input_ids, labels))
+    sampler = RandomSampler(data)
+    dataloader = DataLoader(data, sampler=sampler, batch_size=batch_size)
+    return dataloader
+
+
+def _data_augmentation(text, labels, augmenter):
+    """
+    Use an augmentation method to expand a training set.
+
+    :param text: list of input text.
+    :param labels: list of corresponding labels.
+    :param augmenter: textattack.augmentation.Augmenter, augmentation scheme.
+
+    :return: augmented_text, augmented_labels. list of (augmented) input text and labels.
+    """
+    aug_text = augmenter.augment_many(text)
+    # flatten augmented examples and duplicate labels
+    flat_aug_text = []
+    flat_aug_labels = []
+    for i, examples in enumerate(aug_text):
+        for aug_ver in examples:
+            flat_aug_text.append(aug_ver)
+            flat_aug_labels.append(labels[i])
+    return flat_aug_text, flat_aug_labels
+
+
+def _generate_adversarial_examples(model, attackCls, dataset):
+    """
+    Create a dataset of adversarial examples based on perturbations of the existing dataset.
+
+    :param model: Model to attack.
+    :param attackCls: class name of attack recipe to run.
+    :param dataset: iterable of (text, label) pairs.
+
+    :return: list of adversarial examples.
+    """
+    attack = attackCls(model)
+    adv_train_text = []
+    for adv_ex in tqdm.tqdm(
+        attack.attack_dataset(dataset), desc="Attack", total=len(dataset)
+    ):
+        adv_train_text.append(adv_ex.perturbed_text())
```

Comment:
shouldn't this check that adv_ex is a `SuccessfulAttackResult` or `MaximizedAttackResult`?

Reply:
Ah, maybe. if `FailedAttackResult`, and add `original_text()`, else add `perturbed_text()`?

Reply:
Yeah that sounds right. Could maybe have boolean options for:
- Whether to include or ignore Failed/skipped attack results in the adv training epochs
- Whether to use `original_text()` or as far as the attack got (`perturbed_text()`) for `FailedAttackResult`s

Reply:
If this was image adversarial training, we'd include the perturbed version of failed attacks, because they are still the 'worst case' input we can come up with under our own constraints. I don't think you'd want to revert to the clean/best-case. Only problem would be if the perturbed text of a failed result was so distorted by repeated attempts that it becomes a bad sample, but from what I can tell the FailedAttackResult contains the most successful attack and wouldn't run into this...